Switches from params pattern to hieradata

logicminds opened the pull request:
This modernizes the module by switching to the hiera data in module pattern.   I have also simplified the code by bringing the classes into a single class.  It seemed easier to read with everything in one class rather than splitting 1-2 resources into another class where I would have had to reference variables for other classes and potentially break compatibility with prior versions.

This will require a major version change since it changes the params interface a little bit. I thought this would be appropriate since the specific module has not been released under Vox's namespace yet.  

I also updated the sensitive configurations and removed some of the legacy code that was no longer needed.

Additionally, for the config file there is now direct control of some of the configuration via hiera data.

A bit more work after this PR is still needed so this won't be released just yet.   Should we should release a minor update first before merging this?

Incidentally Fixes #12 and #10 because hiera data does not rely on a version string anymore.

logicminds committed:
Moves the params pattern to the modern hiera data lookup system

 * Replaces params with hiera data
 * Simplifies the code by removing the install and config classes

kenyon commented on data/osfamily/freebsd.yaml:
Filename should be `FreeBSD.yaml`

kenyon replied:
Needs to be mixed case to match the case of the fact, `FreeBSD.yaml`.

kenyon commented on data/common.yaml:
These should be moved to `init.pp` so that puppet-strings documents them in `REFERENCE.md`.

kenyon commented on .sync.yml:
missing final newline

kenyon commented on manifests/init.pp:
```diff
@@ -47,25 +47,108 @@
-) inherits autosign::params {
-  contain autosign::install
-  contain autosign::config
+  String               $ensure,
```
```suggestion
  Enum['present', 'absent'] $ensure = 'present',
```

logicminds replied:
I did have an enum but the tests starting failing because the ensure is sometimes used for versions

kenyon commented on manifests/init.pp:
```diff
@@ -47,25 +47,108 @@
+    }
+  }
+
+  $dir_ensure = $autosign::ensure ? {
+    /(absent|purged)/ => 'absent',
+    default           => 'directory',
+  }
+
+  $config_ensure = $autosign::ensure ? {
+    /(absent|purged)/ => 'absent',
+    default           => 'file',
+  }
```
stdlib has a function for this: https://github.com/puppetlabs/puppetlabs-stdlib/blob/main/functions/ensure.pp

logicminds replied:
wish I could use this but it won't work on version strings.

logicminds committed:
Move defaults to code

kenyon commented:
I use `%{facts.is_pe}.yaml` in my control repo `hiera.yaml`, so IMO that is fine.

kenyon commented on data/common.yaml:
`common.yaml` should be deleted. The defaults are already documented as class parameters, so we don't want them repeated as comments. The default for `autosign::configfile` is already in `init.pp`, and we can eliminate `autosign::base_configpath`.

kenyon commented on data/puppet_version/puppet_enterprise/true.yaml:
```diff
@@ -0,0 +1,7 @@
+autosign::group: 'pe-puppet'
+autosign::logpath: '/var/log/puppetlabs/puppetserver'
+autosign::journalpath: '/opt/puppetlabs/server/autosign'
+autosign::base_configpath: '/etc/puppetlabs/puppetserver'
```
```suggestion
autosign::configfile: '/etc/puppetlabs/puppetserver/autosign.conf'
```

kenyon commented on data/osfamily/Suse.yaml:
Delete, since these are defaults.

kenyon commented on data/osfamily/RedHat.yaml:
Delete, since these are defaults.

kenyon commented on data/osfamily/Debian.yaml:
Delete, since these are defaults.

kenyon commented on data/osfamily/freebsd.yaml:
```diff
@@ -0,0 +1,5 @@
+---
+# OS family defaults for FreeBSD and OpenBSD
+autosign::base_configpath: '/usr/local/etc'
```
```suggestion
autosign::configfile: '/usr/local/etc/autosign.conf'
```

kenyon commented on data/puppet_version/puppet_enterprise/true.yaml:
Personally I would get rid of this PE data file too. My position is that only software that actually comes with PE should be putting files in PE directories.

logicminds replied:
Having implemented autosign at several enterprises with PE, I know removing could potentially break folks stuff since the logic was in params prior to this.  Not that they couldn't override hieradata easily but I would be removing backwards compatibility.

kenyon commented on manifests/init.pp:
```diff
@@ -47,25 +47,108 @@
+  String               $log_level            = 'INFO',
+  Integer              $jwt_token_validity   = 7200,
+  Sensitive[String]     $jwt_token_secret    = Sensitive(fqdn_rand_string(30)),
+  Variant[Sensitive[Hash], Hash] $config     = {},
```
We should use stricter types. I'd also eliminate the alignment, since it's tedious to maintain and causes larger diffs when things need to be realigned.

```suggestion
  String[1] $ensure = 'present',
  String[1] $package_name = 'autosign',
  String[1] $puppetserver_ensure = 'present',
  Boolean $manage_journalfile = true,
  Boolean $manage_logfile = true,
  Boolean $manage_package = true,
  Stdlib::Absolutepath $journalpath = '/var/lib/autosign',
  Stdlib::Absolutepath $configfile = '/etc/autosign.conf',
  Stdlib::Absolutepath $logpath = '/var/log',
  String[1] $user = 'puppet',
  String[1] $group = 'puppet',
  String[1] $gem_provider = 'puppet_gem',
  Optional[String[1]] $gem_source = undef,
  String[1] $log_level = 'INFO',
  Integer[1] $jwt_token_validity = 7200,
  Sensitive[String[1]] $jwt_token_secret = Sensitive(fqdn_rand_string(30)),
  Variant[Sensitive[Hash], Hash] $config = {},
```

kenyon commented on manifests/init.pp:
```diff
@@ -47,25 +47,108 @@
+  }
+
+  # Merge the default config with the user-provided config, unwrap before merging 
+  $settings             = Sensitive.new(deep_merge({
```
```suggestion
  $settings = Sensitive.new(deep_merge({
```

kenyon commented on manifests/init.pp:
```diff
@@ -47,25 +47,108 @@
+    'jwt_token' => {
+      'validity'    => $autosign::jwt_token_validity,
+      'journalfile' => "${autosign::journalpath}/autosign.journal",
+      # THIS IS NOT SECURE! It is marginally better than harcoding a password,
+      # but it can be replicated externaly to the Puppet Master.
+      # Please override this. It will also cause multi-master setups to not work
+      # correctly, all the more reason to override it.
```
This kind of comment should not be needed. The code should instead raise an error to avoid an insecure configuration in the first place.

logicminds replied:
valid.  However, this is outside the scope of the PR.   Created #59 to fix this later

kenyon commented on manifests/init.pp:
```diff
@@ -47,25 +47,108 @@
+
+  $sensitive_config = Sensitive(epp('autosign/autosign.conf.epp', { settings => $settings.unwrap }))
+
+  # Ensure we set the value to Sensitive so the secrets don't get revealed
```
Comment seems out of place, and redundant anyway.

```suggestion
```

kenyon commented on manifests/init.pp:
```diff
@@ -47,25 +47,108 @@
+      group  => $autosign::group,
+    }
+  }
+  # the autosign key journal stores previously-used tokens to prevent re-use
```
The purpose of the journal should be documented in the class parameter docs.

kenyon commented on manifests/init.pp:
```diff
@@ -47,25 +47,108 @@
+      mode    => '0640',
+      owner   => $autosign::user,
+      group   => $autosign::group,
+      require => File[$autosign::journalpath],
```
Parent directories are autorequired.

```suggestion
      ensure => 'file',
      mode   => '0640',
      owner  => $autosign::user,
      group  => $autosign::group,
```

logicminds committed:
Refactor hiera data structure and consolidate configuration

Simplifies the hiera data hierarchy by removing redundant OS-specific
configuration files and adding explicit configfile parameters where needed.

Changes:
- Removed data/common.yaml (consolidated into platform-specific files)
- Removed data/osfamily/Debian.yaml (redundant configuration)
- Removed data/osfamily/RedHat.yaml (redundant configuration)
- Removed data/osfamily/Suse.yaml (redundant configuration)
- Added explicit configfile parameter to FreeBSD hiera data
- Added explicit configfile parameter to Puppet Enterprise hiera data
- Minor formatting cleanup in manifests/init.pp

logicminds committed:
Add stricter types for string parameters

  * Remove extra spacing in parameters

logicminds committed:
Improve documentation and code cleanup in autosign class

- Add documentation for journalpath and jwt_token_secret parameters
- Fix typo: 'Weather' -> 'Whether' in manage_journalfile parameter
- Update jwt_token_secret type constraint to String[30] and increase random string length to 32
- Remove redundant comments and unnecessary require dependency

bastelfreak commented on Gemfile:
```diff
@@ -5,8 +5,8 @@ source ENV['GEM_SOURCE'] || 'https://rubygems.org'
-  gem 'puppet_metadata', '~> 6.0',  :require => false
-  gem 'autosign',                   :require => false
+  gem 'autosign', :require => false
+  gem 'puppet_metadata', '~> 5.0',  :require => false
```
Why the downgrade?

logicminds replied:
good catch.  Bad rebase and regeneration of the Gemfile I guess.
